File: src/cmds/gain.ts
import { TokConfig } from '../core/config';
import { DB, readCommands, CommandRow } from '../core/local-db';
import { dollar, estimateTokens, percent, formatNumber, withinDays } from '../core/utils';

interface GainArgs {
  graph?: boolean;
  history?: boolean;
  daily?: boolean;
  format?: string;
}

interface DayBucket {
  day: string;
  inBytes: number;
  outBytes: number;
  saved: number;
  count: number;
}

export function runGain(db: DB, config: TokConfig, args: GainArgs): string {
  if (args.format === 'json') return jsonExport(db);
  if (args.graph) return graphView(db);
  if (args.history) return historyView(db);
  if (args.daily) return dailyView(db);
  return summaryView(db, config);
}

function summaryView(db: DB, config: TokConfig): string {
  const cmds = readCommands(db);
  const today = aggregate(cmds, 1);
  const week = aggregate(cmds, 7);
  const allTime = aggregate(cmds, 36500);

  const lines: string[] = [];
  lines.push('tok savings - filter compression');
  lines.push('═'.repeat(58));
  lines.push(`Today:        ${tokRow(today, config)}`);
  lines.push(`Last 7 days:  ${tokRow(week, config)}`);
  lines.push(`All time:     ${tokRow(allTime, config)}`);

  lines.push('');
  lines.push('Top commands today:');
  const todayGroups = new Map<string, { runs: number; sumPct: number }>();
  for (const r of cmds) {
    if (!withinDays(r.timestamp, 1)) continue;
    const g = todayGroups.get(r.cmd_type) || { runs: 0, sumPct: 0 };
    g.runs += 1;
    g.sumPct += r.savings_pct;
    todayGroups.set(r.cmd_type, g);
  }
  const topCmds = Array.from(todayGroups.entries())
    .map(([cmd_type, g]) => ({ cmd_type, runs: g.runs, avgPct: g.runs ? g.sumPct / g.runs : 0 }))
    .sort((a, b) => b.runs - a.runs)
    .slice(0, 8);
  if (topCmds.length === 0) {
    lines.push('  (no data yet)');
  } else {
    for (const row of topCmds) {
      lines.push(`  ${row.cmd_type.padEnd(14)} ${percent(row.avgPct, 0).padEnd(5)} ${row.runs} runs`);
    }
  }

  const unoptCounts = new Map<string, number>();
  for (const r of cmds) {
    if (r.savings_pct < 1 && withinDays(r.timestamp, 7)) {
      unoptCounts.set(r.cmd_type, (unoptCounts.get(r.cmd_type) || 0) + 1);
    }
  }
  const unopt = Array.from(unoptCounts.entries())
    .map(([cmd_type, runs]) => ({ cmd_type, runs }))
    .sort((a, b) => b.runs - a.runs)
    .slice(0, 5);
  if (unopt.length > 0) {
    lines.push('');
    lines.push('Not yet optimized (0% savings):');
    const compact = unopt.map((r) => `${r.cmd_type} (×${r.runs})`).join('  ');
    lines.push(`  ${compact}`);
    lines.push('  → Run: tok discover for optimization suggestions');
  }
  return lines.join('\n');
}

function tokRow(b: DayBucket, config: TokConfig): string {
  const inT = estimateTokens(' '.repeat(b.inBytes));
  const outT = estimateTokens(' '.repeat(b.outBytes));
  const savedT = estimateTokens(' '.repeat(b.saved));
  const pct = b.inBytes > 0 ? (b.saved / b.inBytes) * 100 : 0;
  const cost = (savedT / 1000) * config.tokenPricePer1k;
  return `${formatNumber(inT)} → ${formatNumber(outT)} tokens   saved ${pct.toFixed(0)}%  (~${dollar(cost)})`;
}

function aggregate(cmds: CommandRow[], days: number): DayBucket {
  let inB = 0;
  let outB = 0;
  let sB = 0;
  let n = 0;
  for (const r of cmds) {
    if (!withinDays(r.timestamp, days)) continue;
    inB += r.input_bytes;
    outB += r.out_bytes;
    sB += r.saved_bytes;
    n += 1;
  }
  return { day: '', inBytes: inB, outBytes: outB, saved: sB, count: n };
}

function graphView(db: DB): string {
  const byDay = new Map<string, number>();
  for (const r of readCommands(db)) {
    if (!withinDays(r.timestamp, 30)) continue;
    const day = r.timestamp.slice(0, 10);
    byDay.set(day, (byDay.get(day) || 0) + r.saved_bytes);
  }
  const rows = Array.from(byDay.entries())
    .map(([day, sB]) => ({ day, sB }))
    .sort((a, b) => (a.day < b.day ? -1 : 1));
  if (rows.length === 0) return 'No data yet.';
  const max = Math.max(...rows.map((r) => r.sB));
  const lines: string[] = ['Daily savings (last 30 days, bytes):', ''];
  for (const r of rows) {
    const len = max > 0 ? Math.round((r.sB / max) * 40) : 0;
    const bar = '█'.repeat(len);
    lines.push(`  ${r.day} ${bar} ${formatNumber(r.sB)}`);
  }
  return lines.join('\n');
}

function historyView(db: DB): string {
  const rows = readCommands(db).slice(-20).reverse();
  if (rows.length === 0) return 'No history yet.';
  const lines: string[] = ['Last 20 commands:', ''];
  for (const r of rows) {
    lines.push(
      `  ${r.timestamp.replace('T', ' ').slice(0, 19)} ${r.cmd_type.padEnd(14)} ${formatNumber(r.input_bytes).padStart(8)} → ${formatNumber(r.out_bytes).padStart(8)}  ${percent(r.savings_pct, 0).padStart(5)}`,
    );
  }
  return lines.join('\n');
}

function dailyView(db: DB): string {
  const byDay = new Map<string, { runs: number; inB: number; outB: number; sB: number }>();
  for (const r of readCommands(db)) {
    if (!withinDays(r.timestamp, 30)) continue;
    const day = r.timestamp.slice(0, 10);
    const g = byDay.get(day) || { runs: 0, inB: 0, outB: 0, sB: 0 };
    g.runs += 1;
    g.inB += r.input_bytes;
    g.outB += r.out_bytes;
    g.sB += r.saved_bytes;
    byDay.set(day, g);
  }
  const rows = Array.from(byDay.entries())
    .map(([day, g]) => ({ day, ...g }))
    .sort((a, b) => (a.day < b.day ? 1 : -1));
  if (rows.length === 0) return 'No data yet.';
  const lines: string[] = [
    'Daily savings (last 30 days)',
    '─'.repeat(70),
    'Day         Runs      Input         Output          Saved      %',
    '─'.repeat(70),
  ];
  for (const r of rows) {
    const pct = r.inB > 0 ? (r.sB / r.inB) * 100 : 0;
    lines.push(
      `${r.day}  ${String(r.runs).padStart(5)} ${formatNumber(r.inB).padStart(11)} ${formatNumber(r.outB).padStart(13)} ${formatNumber(r.sB).padStart(13)} ${percent(pct, 0).padStart(5)}`,
    );
  }
  return lines.join('\n');
}

function jsonExport(db: DB): string {
  const rows = readCommands(db).slice().reverse();
  return JSON.stringify({ rows }, null, 2);
}

File: internal/analytics/gain.go
package analytics

import (
	"fmt"
	"sort"
	"strconv"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/config"
	"github.com/Kevalgor12/tok-proxy/internal/store"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

type GainArgs struct {
	Graph   bool
	History bool
	Daily   bool
	Format  string
}

type dayBucket struct {
	inBytes  int
	outBytes int
	saved    int
	count    int
}

// RunGain reports filter-compression savings. Views: summary (default), graph, history,
// daily table, and a JSON export.
func RunGain(s *store.Store, cfg config.Config, args GainArgs) string {
	switch {
	case args.Format == "json":
		return gainJSONExport(s)
	case args.Graph:
		return gainGraphView(s)
	case args.History:
		return gainHistoryView(s)
	case args.Daily:
		return gainDailyView(s)
	default:
		return gainSummaryView(s, cfg)
	}
}

func gainAggregate(cmds []store.CommandRow, days int) dayBucket {
	var b dayBucket
	for _, r := range cmds {
		if !util.WithinDays(r.Timestamp, days) {
			continue
		}
		b.inBytes += r.InputBytes
		b.outBytes += r.OutBytes
		b.saved += r.SavedBytes
		b.count++
	}
	return b
}

func gainTokRow(b dayBucket, cfg config.Config) string {
	inT := util.BytesToTokens(b.inBytes)
	outT := util.BytesToTokens(b.outBytes)
	savedT := util.BytesToTokens(b.saved)
	pct := 0.0
	if b.inBytes > 0 {
		pct = float64(b.saved) / float64(b.inBytes) * 100
	}
	cost := float64(savedT) / 1000 * cfg.TokenPricePer1k
	return fmt.Sprintf("%s → %s tokens   saved %.0f%%  (~%s)",
		util.FormatNumber(inT), util.FormatNumber(outT), pct, util.Dollar(cost))
}

func gainSummaryView(s *store.Store, cfg config.Config) string {
	cmds := s.ReadCommands()

	lines := []string{
		"tok savings - filter compression",
		strings.Repeat("═", 58),
		"Today:        " + gainTokRow(gainAggregate(cmds, 1), cfg),
		"Last 7 days:  " + gainTokRow(gainAggregate(cmds, 7), cfg),
		"All time:     " + gainTokRow(gainAggregate(cmds, 36500), cfg),
		"",
		"Top commands today:",
	}

	type todayGroup struct {
		runs   int
		sumPct float64
	}
	groups := newGroups[todayGroup]()
	for _, r := range cmds {
		if !util.WithinDays(r.Timestamp, 1) {
			continue
		}
		g := groups.get(r.CmdType)
		g.runs++
		g.sumPct += r.SavingsPct
	}
	type topRow struct {
		cmdType string
		runs    int
		avgPct  float64
	}
	var topCmds []topRow
	for _, e := range groups.entries() {
		avg := 0.0
		if e.val.runs > 0 {
			avg = e.val.sumPct / float64(e.val.runs)
		}
		topCmds = append(topCmds, topRow{e.key, e.val.runs, avg})
	}
	sort.SliceStable(topCmds, func(i, j int) bool { return topCmds[i].runs > topCmds[j].runs })
	topCmds = head(topCmds, 8)
	if len(topCmds) == 0 {
		lines = append(lines, "  (no data yet)")
	} else {
		for _, row := range topCmds {
			lines = append(lines, fmt.Sprintf("  %s %s %d runs",
				util.Pad(row.cmdType, 14, false), util.Pad(util.Percent(row.avgPct, 0), 5, false), row.runs))
		}
	}

	unopt := newGroups[int]()
	for _, r := range cmds {
		if r.SavingsPct < 1 && util.WithinDays(r.Timestamp, 7) {
			*unopt.get(r.CmdType) += 1
		}
	}
	type unoptRow struct {
		cmdType string
		runs    int
	}
	var unoptRows []unoptRow
	for _, e := range unopt.entries() {
		unoptRows = append(unoptRows, unoptRow{e.key, *e.val})
	}
	sort.SliceStable(unoptRows, func(i, j int) bool { return unoptRows[i].runs > unoptRows[j].runs })
	unoptRows = head(unoptRows, 5)
	if len(unoptRows) > 0 {
		var compact []string
		for _, r := range unoptRows {
			compact = append(compact, fmt.Sprintf("%s (×%d)", r.cmdType, r.runs))
		}
		lines = append(lines,
			"",
			"Not yet optimized (0% savings):",
			"  "+strings.Join(compact, "  "),
			"  → Run: tok discover for optimization suggestions")
	}
	return strings.Join(lines, "\n")
}

func gainGraphView(s *store.Store) string {
	byDay := newGroups[int]()
	for _, r := range s.ReadCommands() {
		if !util.WithinDays(r.Timestamp, 30) {
			continue
		}
		*byDay.get(day10(r.Timestamp)) += r.SavedBytes
	}
	type row struct {
		day string
		sB  int
	}
	var rows []row
	for _, e := range byDay.entries() {
		rows = append(rows, row{e.key, *e.val})
	}
	if len(rows) == 0 {
		return "No data yet."
	}
	sort.SliceStable(rows, func(i, j int) bool { return rows[i].day < rows[j].day })
	vals := make([]int, len(rows))
	for i, r := range rows {
		vals[i] = r.sB
	}
	max := maxInt(vals)
	lines := []string{"Daily savings (last 30 days, bytes):", ""}
	for _, r := range rows {
		lines = append(lines, fmt.Sprintf("  %s %s %s", r.day, bar(r.sB, max, 40), util.FormatNumber(r.sB)))
	}
	return strings.Join(lines, "\n")
}

func gainHistoryView(s *store.Store) string {
	rows := reversed(tail(s.ReadCommands(), 20))
	if len(rows) == 0 {
		return "No history yet."
	}
	lines := []string{"Last 20 commands:", ""}
	for _, r := range rows {
		ts := strings.ReplaceAll(r.Timestamp, "T", " ")
		if len(ts) > 19 {
			ts = ts[:19]
		}
		lines = append(lines, fmt.Sprintf("  %s %s %s → %s  %s",
			ts,
			util.Pad(r.CmdType, 14, false),
			util.Pad(util.FormatNumber(r.InputBytes), 8, true),
			util.Pad(util.FormatNumber(r.OutBytes), 8, true),
			util.Pad(util.Percent(r.SavingsPct, 0), 5, true)))
	}
	return strings.Join(lines, "\n")
}

func gainDailyView(s *store.Store) string {
	type g struct{ runs, inB, outB, sB int }
	byDay := newGroups[g]()
	for _, r := range s.ReadCommands() {
		if !util.WithinDays(r.Timestamp, 30) {
			continue
		}
		x := byDay.get(day10(r.Timestamp))
		x.runs++
		x.inB += r.InputBytes
		x.outB += r.OutBytes
		x.sB += r.SavedBytes
	}
	type row struct {
		day                 string
		runs, inB, outB, sB int
	}
	var rows []row
	for _, e := range byDay.entries() {
		rows = append(rows, row{e.key, e.val.runs, e.val.inB, e.val.outB, e.val.sB})
	}
	if len(rows) == 0 {
		return "No data yet."
	}
	sort.SliceStable(rows, func(i, j int) bool { return rows[i].day > rows[j].day })
	lines := []string{
		"Daily savings (last 30 days)",
		strings.Repeat("─", 70),
		"Day         Runs      Input         Output          Saved      %",
		strings.Repeat("─", 70),
	}
	for _, r := range rows {
		pct := 0.0
		if r.inB > 0 {
			pct = float64(r.sB) / float64(r.inB) * 100
		}
		lines = append(lines, fmt.Sprintf("%s  %s %s %s %s %s",
			r.day,
			util.Pad(strconv.Itoa(r.runs), 5, true),
			util.Pad(util.FormatNumber(r.inB), 11, true),
			util.Pad(util.FormatNumber(r.outB), 13, true),
			util.Pad(util.FormatNumber(r.sB), 13, true),
			util.Pad(util.Percent(pct, 0), 5, true)))
	}
	return strings.Join(lines, "\n")
}

func gainJSONExport(s *store.Store) string {
	return jsonPretty(map[string]any{"rows": reversed(s.ReadCommands())})
}
